Python/fileutils.c

```diff
@@ -1804,6 +1812,41 @@ _Py_write_impl(int fd, const void *buf, size_t count, int gil_held)
             }
         }
     }
+
+    // In non-blocking mode, a write that exceeds the available size of a pipe
+    // fails without writing anything. Limiting writes to the pipe size allows
+    // a buffered write to succeed eventually, as the pipe is read.
+    HANDLE hfile = _Py_get_osfhandle(fd);
+    if (hfile == INVALID_HANDLE_VALUE) {
+        return -1;
+    }
+    DWORD mode, pipe_size;
+    if (GetFileType(hfile) == FILE_TYPE_PIPE &&
+        GetNamedPipeHandleStateW(hfile, &mode,
+            NULL, NULL, NULL, NULL, 0) &&
+        mode & PIPE_NOWAIT)
+    {
+        if (gil_held) {
+            Py_BEGIN_ALLOW_THREADS
+            // GetNamedPipeInfo() requires FILE_READ_ATTRIBUTES access.
+            // CreatePipe() includes this access for the write handle.
+            if (!GetNamedPipeInfo(hfile, NULL, NULL, &pipe_size, NULL)) {
+                pipe_size = 4096;
+            }
+            if (count > pipe_size) {
+                count = pipe_size;
+            }
+            Py_END_ALLOW_THREADS
+        } else {
+            if (!GetNamedPipeInfo(hfile, NULL, NULL, &pipe_size, NULL)) {
+                pipe_size = 4096;
+            }
+            if (count > pipe_size) {
+                count = pipe_size;
+            }
+        }
+    }
```

Comment:
We have to release the GIL when calling `GetFiletype()` and `GetNamedPipeHandleStateW()`.

```suggestion
    if (gil_held) {
        Py_BEGIN_ALLOW_THREADS
        if (GetFileType(hfile) == FILE_TYPE_PIPE &&
            GetNamedPipeHandleStateW(hfile, &mode, NULL, NULL, NULL,
                                     NULL, 0) &&
            mode & PIPE_NOWAIT)
        {
            // GetNamedPipeInfo() requires FILE_READ_ATTRIBUTES access.
            // CreatePipe() includes this access for the write handle.
            if (!GetNamedPipeInfo(hfile, NULL, NULL, &pipe_size, NULL)) {
                pipe_size = 4096;
            }
            if (count > pipe_size) {
                count = pipe_size;
            }
        }
        Py_END_ALLOW_THREADS
    }
    else {
        if (GetFileType(hfile) == FILE_TYPE_PIPE &&
            GetNamedPipeHandleStateW(hfile, &mode, NULL, NULL, NULL,
                                     NULL, 0) &&
            mode & PIPE_NOWAIT)
        {
            // GetNamedPipeInfo() requires FILE_READ_ATTRIBUTES access.
            // CreatePipe() includes this access for the write handle.
            if (!GetNamedPipeInfo(hfile, NULL, NULL, &pipe_size, NULL)) {
                pipe_size = 4096;
            }
            if (count > pipe_size) {
                count = pipe_size;
            }
        }
    }
```
---

`GetNamedPipeHandleStateW()` can block because it's a synchronous operation that has to acquire the file lock in the kernel. `GetFiletype()` is normally fast and doesn't block as long as the real device object is directly accessible from the kernel file object. However, even `GetFileType()` can block if the file is opened on the MUP device (multiple UNC provider) and redirected through a provider such as SMB. For example, "//./pipe/pipename" is not redirected since "//./" is the local device namespace. OTOH, "//server/pipe/pipename" opens on the MUP device and is redirected through the SMB provider, even if "server" is the local machine name.